### PR TITLE
separate logic from config / refactor

### PR DESCRIPTION
**Pull request by nikitonskii**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,13 +1,13 @@
 import React from 'react';
 import logo from './logo.svg';
 import './App.scss';
-import SighnIn from './pages/sighnIn/SighnIn';
+import SignIn from './pages/signIn/SignIn';
 
 function App() {
```
Comment (twixDuo) [inline src/App.tsx]: change to arrow function

File: src/App.tsx [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
   return (
-    <div className="App">
```
Comment (twixDuo) [inline src/App.tsx:8]: add to config el lintera ' => '' and add the` ;` on then end

File: src/components/Button/types.ts [hunk 3]
```diff
@@ -1,4 +1,4 @@
-export interface IProps {
+export interface IButtonProps {
   title: string
   onAction?: React.MouseEventHandler
   buttonType?: 'submit' | 'reset' | 'button'
```
Comment (twixDuo) [inline src/components/Button/types.ts]: its should be enum type

File: src/components/Button/types.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-export interface IProps {
+export interface IButtonProps {
   title: string
   onAction?: React.MouseEventHandler
```
Comment (twixDuo) [inline src/components/Button/types.ts]: its a function ? why you write  React.MouseEventHandler ? its type of event. 

its must be `function` or () => ...

File: src/components/Input/Input.tsx [hunk 5]
```diff
@@ -1,34 +1,33 @@
...
+}): JSX.Element => {
+  
   return (
-    <div className='InputContainer'>
-      <label className='InputLabel' htmlFor={label}>{label}</label>
+    <div className='input-container'>
+      <label className='input-label' htmlFor={label}>{label}</label>
       <input
-        className={error ? 'InputError' : 'Input'}
+        className={isError ? 'input-error' : 'input'}
```
Comment (twixDuo) [inline src/components/Input/Input.tsx]: you should do that =>  className={isError ? 'input error' : 'input'}. you have your head name of class `input` and then you add class `error` if you wont add something new

File: src/components/Input/Input.tsx [hunk 6]
```diff
@@ -1,34 +1,33 @@
...
         type={type}
         id={id}
-        required={!!required}
+        required={!!isRequired}
         onChange={(e: any) => onAction(e)}
```
Comment (twixDuo) [inline src/components/Input/Input.tsx]: short note onChange={onAction}

File: src/components/Input/Input.tsx [hunk 7]
```diff
@@ -1,34 +1,33 @@
...
-      <label className='InputLabel' htmlFor={label}>{label}</label>
+    <div className='input-container'>
+      <label className='input-label' htmlFor={label}>{label}</label>
       <input
-        className={error ? 'InputError' : 'Input'}
+        className={isError ? 'input-error' : 'input'}
         type={type}
         id={id}
-        required={!!required}
+        required={!!isRequired}
```
Comment (twixDuo) [inline src/components/Input/Input.tsx]: isRequired - boolean type, but I see that you change type to boolean when use !!. why you do this ?

File: src/components/Input/Input.tsx [hunk 8]
```diff
@@ -1,34 +1,33 @@
...
         onChange={(e: any) => onAction(e)}
         placeholder={placeholder}
       />
-      {leftIcon ? <span className='InputIconContainer' onClick={(e: any) => onIconAction(e)}><i className={`${leftIcon} InputIcon`} ></i></span> : null}
-      {error ? <p className='InputErrorText'>{errorText}</p> : null}
+      {leftIcon ? <span className='input-icon-container' onClick={(e: any) => onIconAction(e)}><i className={`${leftIcon} input-icon`} ></i></span> : null}
```
Comment (twixDuo) [inline src/components/Input/Input.tsx]: I dont understand  what is leftIcon ? can you change name this icon ?

Comment (twixDuo) [reply src/components/Input/Input.tsx]: short note onClick={onIconAction}

Comment (twixDuo) [reply src/components/Input/Input.tsx]: {leftIcon && (
<span className='input-icon-container' onClick={(e: any) => onIconAction(e)}>
<i className={${leftIcon} input-icon} >
)}

change ? to && and you need write null

File: src/pages/signIn/SignIn.tsx [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  const [formFields, setFormFields] = useState<Fields>({})
+  const [isFieldsValid, setIsFieldsValid] = useState<ValidationFields>({})
+  const [isShowPassword, setShowPassword] = useState<Boolean>(true)
+
+  useLayoutEffect(() => {
+    let fields: Fields = {}
+    let validationVars: ValidationFields = {}
+    config.forEach((field) => {
+      const {id} = field
+      if(field.isRequired) {
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx]: you can write

fields[id]: '';

if(field.isRequired) {
validationVars[id]: false
}

Comment (twixDuo) [reply src/pages/signIn/SignIn.tsx]: and remove duplicate code

File: src/pages/signIn/SignIn.tsx [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+          ...fields,
+          [id]: '',
+        }
+      }
+    })
+    setFormFields(fields)
+    setIsFieldsValid(validationVars)
+  }, [])
+
+  const onChangeFieldsValue = (e: any) => {
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx:39]: name onChangeFieldValue you change value for one field when you run this function

File: src/pages/signIn/SignIn.tsx [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    const {id, value} = e.target
+    setFormFields({
+      ...formFields,
+      [id]: value,
+    })
+  }
+
+  const showPassword = () => {
+    setShowPassword(!isShowPassword)
+  }
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx]: const showPassword = () => setShowPassword(!isShowPassword)

File: src/pages/signIn/SignIn.tsx [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+  const checkFields = (config: ConfigItem[]) => {
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx]: function return boolean type ? you should rename to isValidationError

File: src/pages/signIn/SignIn.tsx [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+          />
+        )}
+          <Button 
+            title='Send'
+            onAction={makeFakeRequest}
+            buttonType='submit'
+          />
+      </form>
+      {success
+        ? <div className='sign-in-popup'>
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx]: success &&

File: src/pages/signIn/types.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
+
+export interface Fields {
+  [propName: string]: any
```
Comment (twixDuo) [inline src/pages/signIn/types.ts]: string

File: src/utils/helpers.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+export function capitalizeFirstLetter(str: string) {
```
Comment (twixDuo) [inline src/utils/helpers.ts]: why not arrow function ? you arrow function if you dont work whit this

File: src/pages/signIn/SignIn.tsx [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+  const onChangeFieldsValue = (e: any) => {
+    const {id, value} = e.target
+    setFormFields({
+      ...formFields,
+      [id]: value,
+    })
+  }
+
+  const showPassword = () => {
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx]: move to component Input

File: src/pages/signIn/SignIn.tsx [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+      if(field.isRequired) {
+        state = {
+          ...state,
+          [field.id]: false,
+        }
+      }
+    })
+    const fields = [...Object.keys(state)]
+    fields.forEach((item) => {
+      if(formFields[item].length) state[item] = false
```
Comment (twixDuo) [inline src/pages/signIn/SignIn.tsx]: state[item] = !!formFields[item].length

File: src/pages/signIn/config.ts [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  {
+    id: 'repeatedPassword',
+    type: 'password',
+    label: 'Repeat Password *',
+    isError: true,
+    isRequired: true,
+    placeholder: 'repeat password',
+    errorText: 'this field must be filled',
+    onIconAction: true,
+    optionalType: 'text',
```
Comment (twixDuo) [inline src/pages/signIn/config.ts]: one type should be, why here i see  optionalType: text ? and in interface you should add `;` on the end

**Review by twixDuo — CHANGES_REQUESTED** [review]

File: src/components/Input/Input.tsx (hunk 7 above)
Comment (nikitonskii) [reply src/components/Input/Input.tsx]: the thing is i used in an interface 'isRequired?: boolean' instead 'isRequired: boolean', have already solved

File: src/components/Input/Input.tsx (hunk 8 above)
Comment (nikitonskii) [reply src/components/Input/Input.tsx]: yep, it was right icon. i confused sides

**Pull request merged**